AttributeError: 'NoneType' object has no attribute 'signal_triggered' on teardown

I have some code which looks like this:

```python
class TestJavascriptEscape:

    def _test_escape(self, text, qtbot, webframe):
        # ...
        with qtbot.waitSignal(webframe.loadFinished, raising=True):
            webframe.setHtml(html_source)
        # ...

    @pytest.mark.parametrize('text', TESTS)
    def test_real_escape(self, webframe, qtbot, text):
        """Test javascript escaping with a real QWebPage."""
        self._test_escape(text, qtbot, webframe)

    @hypothesis.given(hypothesis.strategies.text())
    def test_real_escape_hypothesis(self, webframe, qtbot, text):
        """Test javascript escaping with a real QWebPage and hypothesis."""
        self._test_escape(text, qtbot, webframe)
```

With these fixtures:

```python
@pytest.fixture(scope='session')
def qnam():
    """Session-wide QNetworkAccessManager."""
    from PyQt5.QtNetwork import QNetworkAccessManager
    nam = QNetworkAccessManager()
    nam.setNetworkAccessible(QNetworkAccessManager.NotAccessible)
    return nam


@pytest.fixture
def webpage(qnam):
    """Get a new QWebPage object."""
    from PyQt5.QtWebKitWidgets import QWebPage

    page = QWebPage()
    page.networkAccessManager().deleteLater()
    page.setNetworkAccessManager(qnam)
    return page


@pytest.fixture
def webframe(webpage):
    """Convenience fixture to get a mainFrame of a QWebPage."""
    return webpage.mainFrame()
```

When I run this (after adding the `qtbot.waitSignal` call) on Ubuntu Trusty (and only there!) with PyQt 5.2.1, I get [this giant log](http://qutebrowser.org:8010/builders/ubuntu-trusty/builds/755/steps/unittests/logs/stdio). The gist of it is this:

```
==================================== ERRORS ====================================
____ ERROR at teardown of TestJavascriptEscape.test_real_escape_hypothesis _____

self = <_pytest.runner.SetupState object at 0x7f991bf86898>
colitem = <Function 'test_real_escape_hypothesis'>

    def _callfinalizers(self, colitem):
        # [...]

.tox/unittests/lib/python3.4/site-packages/_pytest/runner.py:356: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.tox/unittests/lib/python3.4/site-packages/_pytest/python.py:1868: in finish
    func()
.tox/unittests/lib/python3.4/site-packages/_pytest/python.py:1826: in teardown
    next()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

qapp = <PyQt5.QtWidgets.QApplication object at 0x7f9913d87af8>
request = <SubRequest 'qtbot' for <Function 'test_real_escape_hypothesis'>>

    @pytest.yield_fixture
    def qtbot(qapp, request):
# [...]
E               Failed: Qt exceptions in virtual methods:
E               ________________________________________________________________________________
E                 File "/home/buildbotx/slaves/slave/ubuntu-trusty/build/.tox/unittests/lib/python3.4/site-packages/pytestqt/plugin.py", line 427, in _quit_loop_by_signal
E                   self.signal_triggered = True
E               
E               AttributeError: 'NoneType' object has no attribute 'signal_triggered'
E               ________________________________________________________________________________
E                 File "/home/buildbotx/slaves/slave/ubuntu-trusty/build/.tox/unittests/lib/python3.4/site-packages/pytestqt/plugin.py", line 427, in _quit_loop_by_signal
E                   self.signal_triggered = True
E               
E               AttributeError: 'NoneType' object has no attribute 'signal_triggered'

[repeated... a lot]
```

Do you have any idea what's happening there :worried:? It seems `self` is `None`... I've seen this in `QObjects` which are already `deleteLater()`'ed, but `SignalBlocker` isn't one...

I'll try to produce a minimal example later.

--- a/pytestqt/_tests/test_wait_signal.py
+++ b/pytestqt/_tests/test_wait_signal.py
@@ -217,9 +217,8 @@ def stop_watch():
     return StopWatch()
 
 
-@pytest.mark.parametrize('multiple, raising',
-                         [(True, True), (True, False), (False, True),
-                          (False, False)])
+@pytest.mark.parametrize('multiple', [True, False])
+@pytest.mark.parametrize('raising', [True, False])
 def test_wait_signals_handles_exceptions(qtbot, multiple, raising):
     """
     Make sure waitSignal handles exceptions correctly.
@@ -241,8 +240,8 @@ def test_wait_signals_handles_exceptions(qtbot, multiple, raising):
             raise TestException
 
 
-@pytest.mark.parametrize('multiple, do_timeout',
-                         [(True, False), (True, True), (False, False), (False, True)])
+@pytest.mark.parametrize('multiple', [True, False])
+@pytest.mark.parametrize('do_timeout', [True, False])
 def test_wait_twice(qtbot, single_shot, multiple, do_timeout):
     """
     https://github.com/pytest-dev/pytest-qt/issues/69

--- a/pytestqt/_tests/test_wait_signal.py
+++ b/pytestqt/_tests/test_wait_signal.py
@@ -239,3 +239,30 @@ def test_wait_signals_handles_exceptions(qtbot, multiple, raising):
     with pytest.raises(TestException):
         with func(arg, timeout=10, raising=raising):
             raise TestException
+
+
+@pytest.mark.parametrize('multiple, do_timeout',
+                         [(True, False), (True, True), (False, False), (False, True)])
+def test_wait_twice(qtbot, single_shot, multiple, do_timeout):
+    """
+    https://github.com/pytest-dev/pytest-qt/issues/69
+    """
+    signaller = Signaller()
+
+    if multiple:
+        func = qtbot.waitSignals
+        arg = [signaller.signal]
+    else:
+        func = qtbot.waitSignal
+        arg = signaller.signal
+
+    if do_timeout:
+        with func(arg, timeout=100):
+            single_shot(signaller.signal, 200)
+        with func(arg, timeout=100):
+            single_shot(signaller.signal, 200)
+    else:
+        with func(arg):
+            signaller.signal.emit()
+        with func(arg):
+            signaller.signal.emit()

--- a/pytestqt/plugin.py
+++ b/pytestqt/plugin.py
@@ -355,6 +355,10 @@ class _AbstractSignalBlocker(object):
         self.timeout = timeout
         self.signal_triggered = False
         self.raising = raising
+        self._timer = QtCore.QTimer()
+        self._timer.setSingleShot(True)
+        if timeout is not None:
+            self._timer.setInterval(timeout)
 
     def wait(self):
         """
@@ -368,11 +372,24 @@ class _AbstractSignalBlocker(object):
         if self.timeout is None and not self._signals:
             raise ValueError("No signals or timeout specified.")
         if self.timeout is not None:
-            QtCore.QTimer.singleShot(self.timeout, self._loop.quit)
+            self._timer.timeout.connect(self._quit_loop_by_timeout)
+            self._timer.start()
         self._loop.exec_()
         if not self.signal_triggered and self.raising:
             raise SignalTimeoutError("Didn't get signal after %sms." %
                                       self.timeout)
+
+    def _quit_loop_by_timeout(self):
+        self._loop.quit()
+        self._cleanup()
+
+    def _cleanup(self):
+        if self.timeout is not None:
+            try:
+                self._timer.timeout.disconnect(self._quit_loop_by_timeout)
+            except (TypeError, RuntimeError):
+                # already disconnected by Qt?
+                pass
 
     def __enter__(self):
         return self
@@ -426,6 +443,16 @@ class SignalBlocker(_AbstractSignalBlocker):
         """
         self.signal_triggered = True
         self._loop.quit()
+        self._cleanup()
+
+    def _cleanup(self):
+        super(SignalBlocker, self)._cleanup()
+        for signal in self._signals:
+            try:
+                signal.disconnect(self._quit_loop_by_signal)
+            except (TypeError, RuntimeError):
+                # already disconnected by Qt?
+                pass
 
 
 class MultiSignalBlocker(_AbstractSignalBlocker):

--- a/pytestqt/plugin.py
+++ b/pytestqt/plugin.py
@@ -355,9 +355,11 @@ class _AbstractSignalBlocker(object):
         self.timeout = timeout
         self.signal_triggered = False
         self.raising = raising
-        self._timer = QtCore.QTimer()
-        self._timer.setSingleShot(True)
-        if timeout is not None:
+        if timeout is None:
+            self._timer = None
+        else:
+            self._timer = QtCore.QTimer()
+            self._timer.setSingleShot(True)
             self._timer.setInterval(timeout)
 
     def wait(self):
@@ -371,7 +373,7 @@ class _AbstractSignalBlocker(object):
             return
         if self.timeout is None and not self._signals:
             raise ValueError("No signals or timeout specified.")
-        if self.timeout is not None:
+        if self._timer is not None:
             self._timer.timeout.connect(self._quit_loop_by_timeout)
             self._timer.start()
         self._loop.exec_()
@@ -384,7 +386,7 @@ class _AbstractSignalBlocker(object):
         self._cleanup()
 
     def _cleanup(self):
-        if self.timeout is not None:
+        if self._timer is not None:
             try:
                 self._timer.timeout.disconnect(self._quit_loop_by_timeout)
             except (TypeError, RuntimeError):

--- a/pytestqt/plugin.py
+++ b/pytestqt/plugin.py
@@ -377,13 +377,13 @@ class _AbstractSignalBlocker(object):
             self._timer.timeout.connect(self._quit_loop_by_timeout)
             self._timer.start()
         self._loop.exec_()
-        self._cleanup()
         if not self.signal_triggered and self.raising:
             raise SignalTimeoutError("Didn't get signal after %sms." %
                                       self.timeout)
 
     def _quit_loop_by_timeout(self):
         self._loop.quit()
+        self._cleanup()
 
     def _cleanup(self):
         if self._timer is not None:
@@ -445,6 +445,7 @@ class SignalBlocker(_AbstractSignalBlocker):
         """
         self.signal_triggered = True
         self._loop.quit()
+        self._cleanup()
 
     def _cleanup(self):
         super(SignalBlocker, self)._cleanup()

--- a/pytestqt/plugin.py
+++ b/pytestqt/plugin.py
@@ -377,13 +377,13 @@ class _AbstractSignalBlocker(object):
             self._timer.timeout.connect(self._quit_loop_by_timeout)
             self._timer.start()
         self._loop.exec_()
+        self._cleanup()
         if not self.signal_triggered and self.raising:
             raise SignalTimeoutError("Didn't get signal after %sms." %
                                       self.timeout)
 
     def _quit_loop_by_timeout(self):
         self._loop.quit()
-        self._cleanup()
 
     def _cleanup(self):
         if self._timer is not None:
@@ -445,7 +445,6 @@ class SignalBlocker(_AbstractSignalBlocker):
         """
         self.signal_triggered = True
         self._loop.quit()
-        self._cleanup()
 
     def _cleanup(self):
         super(SignalBlocker, self)._cleanup()

